[Bug]: Ubuntu UT CI fails when go-ut-analysis proxy returns 502

### Is there an existing issue for the same bug?

- [x] I have checked the existing issues.

### Branch Name

`main` (observed in PR #26571 CI at head `4cbfa08ff8de0e3f2f7833c270e4a0b25abaf479`)

### Commit ID

`4cbfa08ff8de0e3f2f7833c270e4a0b25abaf479`

### Other Environment Information

- Hardware parameters: GitHub Actions hosted/self-hosted Ubuntu x86 runner used by MatrixOne ALL CI
- OS type: Ubuntu/x86
- Workflow: MatrixOne ALL CI
- Run: https://github.com/matrixorigin/matrixone/actions/runs/30708753033
- Failed job: https://github.com/matrixorigin/matrixone/actions/runs/30708753033/job/91392353495

### Actual Behavior

All Go unit tests completed successfully (`go-test=0`), but the `Unit Testing` step failed afterward while installing the UT diagnostic tool:

```text
go: github.com/matrixorigin/go-ut-analysis@latest: github.com/matrixorigin/go-ut-analysis@v0.0.0-20250711025253-f31acb12d3b1: verifying module: github.com/matrixorigin/go-ut-analysis@v0.0.0-20250711025253-f31acb12d3b1: reading https://proxy.golang.com.cn/sumdb/sum.golang.org/tile/8/2/002: 502 Bad Gateway
server response: error code: 502
[ut_summary] [ERR] failed to install go-ut-analysis
[ut_summary] [ERR] UT diagnostics: go-test=0, analysis=1
make: *** [Makefile:383: ut] Error 1
```

The job therefore reports a UT failure even though the test process itself passed.

### Expected Behavior

A transient failure from the configured Go module proxy should not turn a successful unit-test run into a false UT failure. The diagnostic tool should be made reliably available (for example through a pinned/preinstalled artifact or a retry-capable dependency-fetch step) while still failing when actual unit tests or analysis results fail.

### Steps to Reproduce

1. Run the `MatrixOne ALL CI` workflow with the Ubuntu/x86 UT job.
2. Allow the unit tests to complete successfully.
3. During UT summary generation, make `https://proxy.golang.com.cn` return HTTP 502 for the checksum database tile needed to verify `github.com/matrixorigin/go-ut-analysis`.
4. Observe that installing `go-ut-analysis` fails and the entire UT job exits with code 2 despite `go-test=0`.

### Additional information

This failure is unrelated to PR #26571: its changed packages (`pkg/frontend` and `pkg/sql/plan`) passed in the failing job, the full Go test phase reported `go-test=0`, and only the post-test analysis tool installation failed. The workflow run was attempt 1.

--- a/optools/run_ut.sh
+++ b/optools/run_ut.sh
@@ -32,6 +32,7 @@ fi
 
 shopt -s expand_aliases
 source ./utilities.sh
+source ./ut_tools.bash
 go version
 
 BUILD_WKSP=$(dirname "$PWD") && cd $BUILD_WKSP
@@ -316,7 +317,7 @@ function ut_summary(){
   # analyzer cannot parse a truncated/interleaved go test JSON stream.
   mkdir -p "${report_path}/failed/outputs"
 
-  if ! go install github.com/matrixorigin/go-ut-analysis@latest; then
+  if ! install_go_ut_analysis; then
     analysis_status=1
     logger "ERR" "failed to install go-ut-analysis"
   else

--- a/optools/ut_tools.bash
+++ b/optools/ut_tools.bash
@@ -1,0 +1,61 @@
+#!/bin/bash
+
+# Copyright 2026 Matrix Origin
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+GO_UT_ANALYSIS_VERSION="v0.0.0-20250711025253-f31acb12d3b1"
+
+function retry_command() {
+    if (( $# < 3 )); then
+        echo "Usage: retry_command MAX_ATTEMPTS DELAY_SECONDS COMMAND [ARG...]" >&2
+        return 2
+    fi
+
+    local max_attempts=$1
+    local delay_seconds=$2
+    shift 2
+
+    if ! [[ "${max_attempts}" =~ ^[1-9][0-9]*$ ]]; then
+        echo "max_attempts must be a positive integer, got '${max_attempts}'" >&2
+        return 2
+    fi
+    if ! [[ "${delay_seconds}" =~ ^[0-9]+$ ]]; then
+        echo "delay_seconds must be a non-negative integer, got '${delay_seconds}'" >&2
+        return 2
+    fi
+
+    local attempt
+    local status
+    for (( attempt = 1; attempt <= max_attempts; attempt++ )); do
+        if "$@"; then
+            return 0
+        else
+            status=$?
+        fi
+
+        if (( attempt == max_attempts )); then
+            return "${status}"
+        fi
+        echo "command failed (attempt ${attempt}/${max_attempts}), retrying in ${delay_seconds}s..." >&2
+        sleep "${delay_seconds}"
+    done
+}
+
+function install_go_ut_analysis() {
+    local max_attempts=${1:-3}
+    local delay_seconds=${2:-5}
+
+    retry_command "${max_attempts}" "${delay_seconds}" \
+        go install "github.com/matrixorigin/go-ut-analysis@${GO_UT_ANALYSIS_VERSION}"
+}

--- a/optools/ut_tools_test.go
+++ b/optools/ut_tools_test.go
@@ -1,0 +1,121 @@
+// Copyright 2026 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package optools
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+const goUTAnalysisModule = "github.com/matrixorigin/go-ut-analysis@v0.0.0-20250711025253-f31acb12d3b1"
+
+func writeMockGo(t *testing.T) string {
+	t.Helper()
+
+	dir := t.TempDir()
+	path := filepath.Join(dir, "go")
+	script := `#!/bin/bash
+count=0
+if [[ -f "${MOCK_GO_COUNTER}" ]]; then count=$(<"${MOCK_GO_COUNTER}"); fi
+count=$((count + 1))
+echo "${count}" > "${MOCK_GO_COUNTER}"
+echo "$*" >> "${MOCK_GO_ARGS}"
+if (( count >= MOCK_GO_SUCCEED_AFTER )); then exit 0; fi
+exit "${MOCK_GO_FAILURE_STATUS}"
+`
+	if err := os.WriteFile(path, []byte(script), 0o755); err != nil {
+		t.Fatal(err)
+	}
+	return dir
+}
+
+func runInstall(t *testing.T, attempts, succeedAfter, failureStatus int) ([]byte, int, string, string) {
+	t.Helper()
+
+	toolsPath, err := filepath.Abs("ut_tools.bash")
+	if err != nil {
+		t.Fatal(err)
+	}
+	counter := filepath.Join(t.TempDir(), "attempts")
+	arguments := filepath.Join(t.TempDir(), "arguments")
+	mockGoDir := writeMockGo(t)
+	cmd := exec.Command("bash", "-c", `source "$1"; install_go_ut_analysis "$2" 0`,
+		"bash", toolsPath, strconv.Itoa(attempts))
+	cmd.Env = append(os.Environ(),
+		"PATH="+mockGoDir+string(os.PathListSeparator)+os.Getenv("PATH"),
+		"MOCK_GO_COUNTER="+counter,
+		"MOCK_GO_ARGS="+arguments,
+		"MOCK_GO_SUCCEED_AFTER="+strconv.Itoa(succeedAfter),
+		"MOCK_GO_FAILURE_STATUS="+strconv.Itoa(failureStatus),
+	)
+	output, err := cmd.CombinedOutput()
+	if err == nil {
+		return output, 0, counter, arguments
+	}
+	if exitError, ok := err.(*exec.ExitError); ok {
+		return output, exitError.ExitCode(), counter, arguments
+	}
+	t.Fatalf("install go-ut-analysis: %v", err)
+	return nil, 0, "", ""
+}
+
+func assertAttempts(t *testing.T, counter, arguments string, expected int) {
+	t.Helper()
+
+	attempts, err := os.ReadFile(counter)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if strings.TrimSpace(string(attempts)) != strconv.Itoa(expected) {
+		t.Fatalf("expected %d attempts, got %q", expected, attempts)
+	}
+	invocations, err := os.ReadFile(arguments)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for lineNumber, invocation := range strings.Split(strings.TrimSpace(string(invocations)), "\n") {
+		if invocation != "install "+goUTAnalysisModule {
+			t.Fatalf("attempt %d used unexpected arguments %q", lineNumber+1, invocation)
+		}
+	}
+}
+
+func TestInstallGoUTAnalysisRetriesTransientFailures(t *testing.T) {
+	output, status, counter, arguments := runInstall(t, 3, 3, 42)
+	if status != 0 {
+		t.Fatalf("install failed with status %d: %s", status, output)
+	}
+	assertAttempts(t, counter, arguments, 3)
+}
+
+func TestInstallGoUTAnalysisDoesNotRetrySuccess(t *testing.T) {
+	output, status, counter, arguments := runInstall(t, 3, 1, 42)
+	if status != 0 {
+		t.Fatalf("install failed with status %d: %s", status, output)
+	}
+	assertAttempts(t, counter, arguments, 1)
+}
+
+func TestInstallGoUTAnalysisPreservesFinalFailure(t *testing.T) {
+	output, status, counter, arguments := runInstall(t, 3, 4, 42)
+	if status != 42 {
+		t.Fatalf("expected status 42, got %d: %s", status, output)
+	}
+	assertAttempts(t, counter, arguments, 3)
+}

